### ISSUE TITLE
`FOR ... IN ZIP ...` and `FOR ... IN ENUMERATE`

### ISSUE THREAD

**Issue by edbrannin** [robotframework/robotframework#1952]
It would be nice if we had a built-in way to [zip](https://docs.python.org/2/library/functions.html#zip)-and-iterate over a group of lists.

For example:

```
${numbers}=    Create List    1    2    3
${names}=    Create List    One    Two    Three
:FOR    ${number}    ${name}    IN ZIP    ${numbers}    ${names}
\    Log    ${number} is spelled ${name}
```

I haven't looked at the Robot parser at all yet, but it seems like this should be on par with `FOR ... IN RANGE` in terms of difficulty.

This would be useful shorthand for test cases where we build several lists in parallel and then want to check them against each other.  I could write an example of what this looks like today with `FOR ... IN RANGE` if you like.

**Comment by pekkaklarck** [comment]
I agree this would be useful in some cases. There is no nice way to iterate over two or more lists currently. Are you @edbrannin interested in implementing this? As you noted, this shouldn't be too complicated to implement and there ought to be enough time to get this into RF 2.9.

If `FOR ... IN ZIP` is added, it would be a nice to add similar `FOR ... IN ENUMERATE` too. That needs a separate issue, though.

**Comment by edbrannin** [comment]
Sure, I can give it a shot.

After a quick grep & README, it looks like I'd have to touch the following files:

- `src/robot/parsing/model.py` (the `ForLoop` class)
- `src/robot/running/keywords.py`(another `ForLoop` class)
- `robot/testdoc.py` (Update `JsonConverter._get_for_loop()`)
    - (May need more tests than listed below?)
- `doc/userguide/src/CreatingTestData/AdvancedFeatures.rst`
- `atest/testdata/running/for.robot` (example tests that use the loops)
- `atest/robot/running/for.robot` (Tests that run the above and make sure the right internal bits happened)
- `AUTHORS.txt` (not to be presumptuous; not sure if you prefer having this as part of a PR or to handle it yourselves)

Did I miss anything?

**Comment by pekkaklarck** [comment]
I initially thought this wouldn't require changes to parsing, but you are right that detecting new `IN ZIP` and `IN ENUMERATE` needs changes there too. It might  be a good idea to change parsing so that it just checks for `IN <whatever>` and lets logic later to check what kind of for loop it actually was. It would be great to get this in to RF 2.9, because the required changes can be too big to be done in a minor release.

Let me know if you need more help with this. We could, for example, chat about this on #robotframework IRC channel.

**Comment by edbrannin** [comment]
Yeah, I was thinking I'd do something like that, maybe having an abstract ForLoop base class and several subclasses to handle the different flavors instead of setting behavior-flags like the current for-in-range code.

What's the timeline for 2.9 look like right now?  I'd like to get started on this tonight or tomorrow evening.

**Comment by pekkaklarck** [comment]
Timeline for 2.9 is still pretty open, but I hope we get it out latest in June. We try to release alpha 1 tomorrow and then continue with issues labeled with alpha 2. Getting also this into alpha 2 is still possible.

**Comment by edbrannin** [comment]
Considering how this will double the number of For-loop types (vanilla, in-range, in-zip, in-enumerate) I'm thinking it would be best if I replaced ForLoop with an abstract base class and 4 subclasses (or possibly a base ForLoop class and 3 specialized subclasses).

Either way, I'm planning to replace all the existing `ForLoop()` constructor calls with a factory method that can pick the right kind of loop based on a dict-lookup of the "IN" clause.

```python
LOOP_TYPES = dict(
    IN=BasicForLoop,
    INRANGE=RangeForLoop,
    INZIP=ZipForLoop,
    INENUMERATE=EnumerateForLoop,
    )
def get_for_loop(*stuff):
    # TODO Find & remove spaces from the "IN" cell, like ForLoop._get_range_and_index()
    return LOOP_TYPES[in_value)(*stuff)
```

This would happen for the ForLoop classes in two places:
* `src/robot/parsing/model.py`
* `src/robot/running/keywords.py`

Sound reasonable so far?  Any naming suggestions?

**Comment by pekkaklarck** [comment]
Sounds good. In parsing another solution would be just using one for-loop that has variables, type, and values as separate attributes. Then we could verify type in running side. This would have the benefit that possible new for-loop types would need no further changes in parsing.

**Comment by edbrannin** [comment]
Good idea!  I'm glad I asked before doing the work.

**Comment by edbrannin** [comment]
[How's this look so far](https://github.com/robotframework/robotframework/compare/master...edbrannin:1952-for-in-zip)?  (Feel free to tell me to stop pestering you or take this to IRC until I'm ready for a PR)

**Comment by pekkaklarck** [comment]
There's now PR #1952 implementing this. I don't have time to look at it right now, but I assign this to RF 2.9 alpha 2 scope to make sure we don't forget it.

**Comment by pekkaklarck** [comment]
PR #1989 is getting better and I'm confident we get this into 2.9. The PR is going to cover also `FOR IN ... ENUMERATE` that is currently issue #1954. What do you think, should we merge these issues together? Updating the title of this issue and closing the other ought to be enough.

**Comment by pekkaklarck** [comment]
PR #1989 has been merged. Thanks for great work @edbrannin! We'll close this issue now, but it would be great if you could review the changes we did to code and especially to docs.
